File: content-solutions/dfs/cave-order.js
function solution(n, tunnels, start) {
  const graph = Array.from({ length: n }, () => []);
  for (const [a, b] of tunnels) {
    graph[a].push(b);
    graph[b].push(a);
  }
  for (const neighbors of graph) neighbors.sort((p, q) => p - q);
  const visited = new Array(n).fill(false);
  const order = [];
  function dfs(v) {
    visited[v] = true;
    order.push(v);
    for (const w of graph[v]) {
      if (!visited[w]) dfs(w);
    }
  }
  dfs(start);
  return order;
}

File: content-solutions/dfs/cave-order.java
import java.util.*;

class Solution {
    private List<List<Integer>> graph;
    private boolean[] visited;
    private final List<Integer> order = new ArrayList<>();

    void dfs(int v) {
        visited[v] = true;
        order.add(v);
        for (int w : graph.get(v)) {
            if (!visited[w]) dfs(w);
        }
    }

    public List<Integer> solution(int n, int[][] tunnels, int start) {
        graph = new ArrayList<>();
        for (int i = 0; i < n; i++) graph.add(new ArrayList<>());
        for (int[] e : tunnels) {
            graph.get(e[0]).add(e[1]);
            graph.get(e[1]).add(e[0]);
        }
        for (List<Integer> neighbors : graph) Collections.sort(neighbors);
        visited = new boolean[n];
        dfs(start);
        return order;
    }
}
